// modern-web-dev-build/dist/gulp/tasks/scripts-typescript.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _abstractTaskLoader = require("../abstractTaskLoader");

var _abstractTaskLoader2 = _interopRequireDefault(_abstractTaskLoader);

var _gulpSourcemaps = require("gulp-sourcemaps");

var _gulpSourcemaps2 = _interopRequireDefault(_gulpSourcemaps);

var _gulpTypescript = require("gulp-typescript");

var _gulpTypescript2 = _interopRequireDefault(_gulpTypescript);

var _gulpSize = require("gulp-size");

var _gulpSize2 = _interopRequireDefault(_gulpSize);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ScriptsTypeScriptTaskLoader = function (_AbstractTaskLoader) {
    _inherits(ScriptsTypeScriptTaskLoader, _AbstractTaskLoader);

    function ScriptsTypeScriptTaskLoader() {
        _classCallCheck(this, ScriptsTypeScriptTaskLoader);

        return _possibleConstructorReturn(this, Object.getPrototypeOf(ScriptsTypeScriptTaskLoader).apply(this, arguments));
    }

    _createClass(ScriptsTypeScriptTaskLoader, [{
        key: "registerTask",
        value: function registerTask(gulp) {
            _get(Object.getPrototypeOf(ScriptsTypeScriptTaskLoader.prototype), "registerTask", this).call(this, gulp);

            gulp.task("scripts-typescript", "Transpile TypeScript to ES, include references to library and app .d.ts files and generate sourcemaps", function () {
                var tsProject = _gulpTypescript2.default.createProject("tsconfig.json", {
                    typescript: require("typescript") });

                var tsConfigOutDir = tsProject.config.compilerOptions.outDir;

                var tsResult = tsProject.src().pipe(_gulpSourcemaps2.default.init()).pipe((0, _gulpTypescript2.default)(tsProject));

                tsResult.dts.pipe(gulp.dest(tsConfigOutDir));

                return tsResult.js.pipe(_gulpSourcemaps2.default.write(".", {
                    includeContent: false,
                    sourceRoot: "/" })).pipe(gulp.dest(tsConfigOutDir)).pipe((0, _gulpSize2.default)({
                    title: "scripts-typescript"
                }));
            });
        }
    }]);

    return ScriptsTypeScriptTaskLoader;
}(_abstractTaskLoader2.default);

module.exports = new ScriptsTypeScriptTaskLoader();
